src/compiler/checker.ts

```diff
@@ -21479,18 +21608,262 @@ namespace ts {
                 return type;
             }
 
+            // return true only when it is a access expression and has ?.
+            function isAccessExpressionContainOptionalChain(e: UnaryExpression) {
+                while (isAccessExpression(e)) {
+                    if (e.flags & NodeFlags.OptionalChain) {
+                        return true;
+                    }
+                    e = e.expression;
+                }
+                return false;
+            }
+
+            /**
+             * @param optionalChainSlice If this is true, only match the part before first optional chain. if expression is a.b.c?.d.e, only take a.b.c to get property.
+             */
+            function getPropertyTypesFromTypeAccordingToExpression(type: Type, expressionWithOutKeyword: Expression, optionalChainSlice = false): Type[] | undefined {
+                // This function is designed to be used in the first steps in follow steps:
+                // 1. get all properties' type of the origional types according to expression path
+                //   - In some case, should not narrow type according to the whole path, for example:
+                //   - typeof a.b?.c.d? !== number, === undefined, !==undefined, when we come to a, its type could only be narrowed by a.b.
+                //   - But for typeof a.b?.c.d? === number, when we come to a, it could be narrowed by the full path.
+                // 2. use expression to mark wanted types as 'true'
+                // 3. use the mark to get narrowed type from origional type.
+
+                function tryGetPropertyPathsOfReferenceFromExpression(expressionOri: Expression, _ref?: Node) {
+                    function getTypeDepthIfMatch(expression: Expression, ref: Node): number {
+                        let result = 0;
+                        let expr = expression;
+                        if (isMatchingReference(ref, expr)) {
+                            return result;
+                        }
+                        while (isAccessExpression(expr)) {
+                            result = result + 1;
+                            expr = expr.expression;
+                            if (isMatchingReference(ref, expr)) {
+                                return result;
+                            }
+                        }
+                        return -1;
+                    }
+
+                    // If expression is a, return []
+                    // If expression is a.b["c"].d(), return ["b","c","d"]
+                    // NOTE: If element expression is not known in compile progress like a.b[f()].d, the result would be undefined
+                    // //NOTE: this function need improvement, ElementAccessExpression argument might could be known in compile time, like "1"+"2", we should check "12" in the path, but how to get the value?
+                    function getPropertyPathsOfAccessExpression(expressionOri: Expression, depth: number): __String[] | undefined {

```

Comment:
We should unnest functions that don’t close over anything from their out scopes as much as possible. This one looks like it doesn’t close over anything at all.